agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -393,5 +397,18 @@ def sync_state_to_master(service_pod, service_container, error = nil)
     def log_service_pod_event(type, data, severity = Logger::INFO)
       super(service_pod.service_id, service_pod.instance_number, type, data, severity)
     end
+
+    # @param container [Docker::Container]
+    # @return [Boolean]
+    def legacy_container?(container)
+      Kontena::ServicePods::Migrator.legacy_container?(container)
+    end
+
+    # @param container [Docker::Container]
+    # @return [Docker::Container]
+    def migrate_container(container)
+      Kontena::ServicePods::Migrator.new(container).migrate
+      container.reload
```

Comment:
...except for this `container.reload`... might be better to do in the `remove_restart_policy`, immediately after the `update`. It's also easier to follow like that.

Reply:
TBH this kind of mutation of the container object's own state is a little nasty in general... you'd like to see something like `service_container = migrate_container(service_container)` to make it clear that the container might change because it got migrated somehow.

Migrations like "the container labels are outdated" would imply that the migrator will terminate the container, and then `ensure_desired_state` would need to continue with `!service_container`. That requires having `migrate_container` return `nil`?